diff --git a/node_modules/expo-modules-core/ios/ReactDelegates/RCTAppDelegate+Recreate.mm b/node_modules/expo-modules-core/ios/ReactDelegates/RCTAppDelegate+Recreate.mm
index 18957b4..8a8b483 100644
--- a/node_modules/expo-modules-core/ios/ReactDelegates/RCTAppDelegate+Recreate.mm
+++ b/node_modules/expo-modules-core/ios/ReactDelegates/RCTAppDelegate+Recreate.mm
@@ -38,6 +38,7 @@ - (UIView *)recreateRootViewWithBundleURL:(nullable NSURL *)bundleURL
     // we don't want to loop the ReactDelegate again. Otherwise it will be infinite loop.
     EXReactRootViewFactory *factory = (EXReactRootViewFactory *)rootViewFactory;
     rootView = [factory superViewWithModuleName:self.moduleName initialProperties:self.initialProps launchOptions:launchOptions];
+    [self customizeRootView:(RCTRootView *)rootView];
   } else {
     rootView = [rootViewFactory viewWithModuleName:self.moduleName initialProperties:self.initialProps launchOptions:launchOptions];
   }
